.travis.yml

```diff
@@ -4,60 +4,13 @@ branches:
 os: osx
 matrix:
   include:
-    # Note: The ordering of the environments is optimized so that the complete matrix runs as quickly as possible,
-    # while also providing the first results as early as possible.
-    # At the moment, build times for each environment are roughly as follows:
-    # xcode7.3: 16 minutes
-    # xcode8.3: 7 minutes
-    # xcode9.2: 7 minutes
-    # xcode10.1: 6 minutes
-    # https://docs.travis-ci.com/user/reference/osx/#macos-version
-    # This data is based on the following build, which should be similar to most builds on the repo:
-    # https://travis-ci.org/pybee/rubicon-objc/builds/506412489
-    # Currently, Travis allows a maximum of two concurrent Mac builds per owner:
-    # https://blog.travis-ci.com/2017-09-22-macos-update
-    # With the following order, xcode7.3 will take up the first slot, and at the same time the three other
-    # environments use the second slot in sequence. xcode10.1 will finish first and provide some early results - most
-    # test failures happen in all environments and will be quickly reported this way.
-
-    # OSX 10.11 El Capitan
-    - osx_image: xcode7.3
     # macOS 10.13 High Sierra
     - osx_image: xcode10.1
-    # macOS 10.13 Mojave
+    # macOS 10.14 Mojave
     - osx_image: xcode10.2
-    # macOS 10.12 Sierra
-    - osx_image: xcode9.2
-    - osx_image: xcode8.3
 language: generic
-cache:
-  directories:
-    ~/.pyenv
-env:
-  global:
-    - PY35="3.5.6"
-    - PY36="3.6.8"
-    - PY37="3.7.2"
 install:
-  # The different Travis environments have different pyenv versions, and not all have the Python versions we need,
-  # so we upgrade pyenv to get a consistent setup in all environments.
-  - brew update
-  - brew outdated pyenv || brew upgrade pyenv
-  # The existing pyenv shims from the build cache may no longer be valid, for example if pyenv was upgraded
-  # (in which case the paths used in the shims point to an older, no longer installed pyenv version).
-  # By rehashing the shims we can be sure that they are always valid again.
-  - pyenv rehash
-  # pyenv's shims directory is not in the PATH by default.
-  - export PATH="$(pyenv root)/shims:${PATH}"
-  # These are the Python versions that we want to test on.
-  # The CFLAGS prefix is to set up the include path so that zlib is available to the Python build
-  # This is needed on Mojave, but doesn't hurt other versions.
-  - CFLAGS="-I$(xcrun --show-sdk-path)/usr/include" pyenv install --skip-existing ${PY35}
-  - CFLAGS="-I$(xcrun --show-sdk-path)/usr/include" pyenv install --skip-existing ${PY36}
-  - CFLAGS="-I$(xcrun --show-sdk-path)/usr/include" pyenv install --skip-existing ${PY37}
-  - pyenv global ${PY37} ${PY36} ${PY35} system
-  - pip install --upgrade tox
+  - python3 -m pip install --upgrade pip setuptools
   - make -f Makefile
 script:
-  - export DYLD_LIBRARY_PATH=`pwd`/tests/objc
-  - tox -e "py{35,36,37}-default,flake8"
```

Comment:
So - I'm not a *huge* fan of tox. 

It played a really useful role when we were trying to build multiple Python versions in a single macOS VM - but outside of that need, I'm not sure I see the value.

* `tox` is itself calling `python setup.py test`, so that's not really an argument either way

* If the Python version being used is a concern, we can invoke using `python3.5`; but even then, I don't see how `tox` provides any more protection than `pyenv` already provides.

* If installability is a concern, then that should be added as part of the test script, regardless how it's being invoked.

The usually given advantages to tox are that it simplifies testing matrix builds. To that, I'd answer that should be the role of the CI system, not a tool you're running locally - if only because any build matrix that's on a single machine will, by definition, be missing a large part of the build matrix. 

tox also obscures success/failure status of individual sub-builds. When you run tox inside a CI build, each tox build can itself have sub-failures, which aren't surfaced to the external build reporting. Or, if they *are* surfaced, it's only because you're using invoking tox in a "build one specific version" configuration - at which point, tox is little more than a configuration language for running your tests... which is what your CI configuration is meant to be.

So - it helps run a *partial* build matrix locally, obscures the collection of test results when used in CI, and adds Yet Another configuration file to your CI configuration. 

My personal preference would be to *remove* tox (and it's configuration).

Reply:
Oh - and the documentation currently doesn't mention tox at all - it's only referenced in the Travis CI configuation. The [contribution docs](https://rubicon-objc.readthedocs.io/en/latest/how-to/contribute.html) reference `python setup.py test`, not `tox` as an entry point.

Reply:
OK, it seems that I've misunderstood tox's intended role in the rubicon-objc tests. Was tox supposed to be just a support tool for the Travis-based CI build and not meant to be used elsewhere? My impression was that tox was intended to be the main entry point for the test suite, for both developers and CI. This also explains:

> the documentation currently doesn't mention tox at all

You're right - the *current* documentation doesn't mention tox. I was accidentally looking at the docs from my big reference documentation PR (#118), where I [changed](https://github.com/beeware/rubicon-objc/commit/dad163e108ff1f453cc8bebb5e2941fd924a99b6#diff-15c2c5daed3ce72b0debfc06a4d9afa6R43) the unit test running documentation to use `tox`, because I thought that was the intended entry point.

I agree that tox's Python version matrix testing is less important if our GitHub Actions build already handles that externally. I care mostly about the other features that tox provides - a central configuration file where you specify how all test/check tools are installed and called, and a single entry point (`tox`) that invokes all of the configured tools.

>     * `tox` is itself calling `python setup.py test`, so that's not really an argument either way

Ah, I should have been clearer with that point. You're right, the current tox.ini still calls `python setup.py test`. But if everything else used tox to run the tests, we would only have to update the single call in the tox.ini to whatever new unit testing tool we decide on, rather than every place in the docs and CI config.

>     * If installability is a concern, then that should be added as part of the test script, regardless how it's being invoked.

The problem is that we don't really have a test script other than tox.ini. We have a CI configuration, but you can't realistically run that locally as a developer. We have a set of unit tests, but those aren't the right place to test installation of the library (rather, the unit tests are run on the already installed library). We have `python setup.py test`, but aside from being deprecated, it just calls the standard `unittest` runner (as I undestand it). Besides, if tox already tests installation into a clean environment by default, I don't see a good reason to reimplement that ourselves.

To be clear, I'm not attached to tox in particular - if you would prefer another tool for configuring/running tests and checkers, I'm open to suggestions. The tox homepage [suggests](https://tox.readthedocs.io/en/latest/#related-projects) [Invoke](https://www.pyinvoke.org/) and [nox](https://nox.thea.codes/en/stable/) as alternatives.

Reply:
It looks like the tox thing is a weird artefact of history. I added it way back in 2014 - as I recall, I was trying it out as a new tool I'd just heard of, and Rubicon's need to test multiple Python versions seemed like a good use case. 

On reflection, I'm not a fan, and I'd honestly rather remove it. I haven't added a tox configuration to any recent project I've built - simply because I don't see that it adds anything. 

I'm honestly not sure I see what the need for a "test script" is. The test script (at present) is `python setup.py test`. That will probably end up being `pytest` in the not-too-distant future. There is a testing prerequisite that you run `make` and set `DYLD_LIBRARY_PATH` - but those can be captured in documentation (especially since they don't need to be run every time you run the test suite during local development). On almost every other project in BeeWare, you only need the single invocation line. It isn't hard to document, though; and honestly, I'd rate the risk of needing to change documentation because we're changing test configuration tool as greater than the risk of needing to change because of the way tests are invoked.

There's also the broader interpretation of "testing" that encompasses linting, checking packaging, checking installation and so on. For me, CI is the ultimate arbiter of all this, because it's the only place where you can actually *run* everything in one hit, in a clean environment that hasn't been affected by the previous test run, etc. Adding a tool that wraps all these features means adding the complexity of yet another thing to configure, where the only real benefit is having a single point of entry to *a* test run - an entry point that, I might add, needs to be documented because no two Python projects use the same entry point. 

And even if we *do* add a single entry point, CI won't be using it anyway, because then you lose the fidelity of reporting that a CI system can provide (i.e., telling you exactly *what* is broken, not just "CI Failed"). So, the "single entry point tool" is reduced to "a complex way to describe what list of commands need to be invoked". For my money, this is much easier to achieve with documentation, backed up with an actual implementation in CI.

So, in summary: I think we should remove the tox configuration; and, to be completely honest, I think we should replace it with "nothing". It's one more thing to maintain that doesn't actually help us guarantee code quality (that's the job of CI), and doesn't provide any significant benefit to onboarding new contributors (not that we're flooded with those anyway).

Reply:
> The test script (at present) is `python setup.py test`. That will probably end up being `pytest` in the not-too-distant future.

Hm, I'm not a huge fan of replacing `python setup.py test` (a generic, test-framework-agnostic entry point) with `pytest` (a specific test runner)... Sadly there is no proper replacement for what `python setup.py test` did (i. e. install the test runner used by the project and use it to run the tests).

The deprecation message from `setup.py test` recommends tox as a replacement, but that recommendation isn't universally agreed upon (see pypa/setuptools#931). I think the general arguments from that issue are similar to some of what you're saying - cross-version testing should be done by CI, and the test runner shouldn't set up its own virtual environments. (And in that context I agree - `python setup.py test` did not set up environments on its own, so its replacement shouldn't either. But that only disqualifies tox as a direct replacement for `setup.py test` - otherwise I think tox is a fine tool and has its uses.)

> There's also the broader interpretation of "testing" that encompasses linting, checking packaging, checking installation and so on. For me, CI is the ultimate arbiter of all this, because it's the only place where you can actually _run_ everything in one hit, in a clean environment that hasn't been affected by the previous test run, etc.

I agree that CI should officially decide whether code is working/acceptable or not, but I think as a dev it should still be possible to easily run tests, linters, etc. locally, without having to commit and push to trigger a CI build. For rubicon-objc my usual development workflow is that while developing I run the unit tests for just one Python version, and once I'm finished and ready to commit/push I run the full tox suite with all supported Python versions and the flake8 check. This way I can catch minor Python version compatibility and formatting problems more quickly than via CI, and it reduces the number of redundant commits.

This is of course still possible without tox - it's just less convenient, as I have to run each step manually (or set my IDE up to do it).

> Adding a tool that wraps all these features means adding the complexity of yet another thing to configure, where the only real benefit is having a single point of entry to _a_ test run - an entry point that, I might add, needs to be documented because no two Python projects use the same entry point.

In general that's true, but tox seems to be fairly widely used - looking at some random high-profile Python projects, [mypy](https://github.com/python/mypy/blob/33fc8a3f308adec1929d8bc09bd1860684c71c69/tox.ini), [pip](https://github.com/pypa/pip/blob/fe3e892a9339f58ee636bcf47427f011a0ae98e1/tox.ini), [setuptools](https://github.com/pypa/setuptools/blob/7e97def47723303fafabe48b22168bbc11bb4821/tox.ini), [numpy](https://github.com/numpy/numpy/blob/ee7fcb790e36d940d4fde69bd1cad5048008064b/tox.ini) and [requests](https://github.com/psf/requests/blob/eedd67462819f8dbf8c1c32e77f9070606605231/tox.ini) all have a tox.ini in their repos. It's fairly likely that people will already know what tox is, and if not, the knowledge will be useful elsewhere.

> And even if we _do_ add a single entry point, CI won't be using it anyway, because then you lose the fidelity of reporting that a CI system can provide (i.e., telling you exactly _what_ is broken, not just "CI Failed").

If you mean that the CI system can no longer differentiate between (for example) the test suite failing and flake8 failing, that could be fixed by invoking the tox environments one at a time (`tox -e py` in one step, `tox -e flake8` in another one, etc.). [mypy's Travis build](https://github.com/python/mypy/blob/33fc8a3f308adec1929d8bc09bd1860684c71c69/.travis.yml) does exactly this - each Travis environment runs one tox environment (possibly with extra arguments). The main advantage of this is that it's relatively easy to replicate each Travis environment locally or on another CI system (except for OS differences obviously) by running tox with the same arguments that were used on Travis.

Reply:
I can understand the desire to have a single common entry point for pre-commit tests; and I agree tox is a relatively common way to do this in the Python ecosystem. 

The root of my concern is when tox *doesn't* meet that requirement. For example, tox *doesn't* currently handle invoking make; so that means we need to communicate that as an additional build requirement. So, we've adopted a tool to provide a single testing entry point... and it doesn't serve as a single testing entry point. I suspect this might be something we can fix with some more tox configuration, but it definitely isn't working like this now.

There are also complications with steps like beefore, which need to be invoked in different ways when they're in CI. Beefore will add comments to the PR to point out problems, but that means it needs to know about the PR environment. That said, this is possibly an argument to drop beefore and move back to a simpler "just invoke flake" configuration for these checks

I guess what I really want is the ability to run Github Actions configurations locally... but I suspect that's a yak I really don't want to start shaving...

Reply:
> For example, tox _doesn't_ currently handle invoking make [...] So, we've adopted a tool to provide a single testing entry point... and it doesn't serve as a single testing entry point. I suspect this might be something we can fix with some more tox configuration, but it definitely isn't working like this now.

Yes, that's not the intent and should be fixed. This should be easy to do, we just need to add the `make` call above `python setup.py test` in the tox.ini.

> There are also complications with steps like beefore, which need to be invoked in different ways when they're in CI. Beefore will add comments to the PR to point out problems, but that means it needs to know about the PR environment.

tox lets you pass command-line arguments into an environment (they are available as `{posargs}` in tox.ini), so `--pull-request` and such can be passed in that way. For the auth token: you can [selectively pass additional envvars from the calling process](https://tox.readthedocs.io/en/latest/config.html#conf-passenv) (by default tox only passes through a few known ones, like `PATH`, and drops everything else to ensure a clean environment).

> I guess what I really want is the ability to run Github Actions configurations locally...

That's sort of the point of tox, except that it's not specific to any CI system. The more configuration you put into tox.ini rather than CI-specific config files, the closer your local tox environment will be to the actual CI environment.

There are limits to this obviously. All of the context information (pull request, etc.) will be missing, but that would always be the case when running locally. You also need to simulate everything by hand that happens before tox is called, but that's mostly environment setup (Git checkout, Python and tox installation) which you usually don't want to replicate exactly (because you already have a fully set up environment). Also you can't call any GitHub Actions actions (is that what they are called?) from tox, but perhaps that's a good thing - it helps minimize vendor lock-in :)